cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +127,191 @@ public Void call() throws Exception {
+                diskFileLocations.remove(fileLocation);
+              }
+              return null;
+            }
+          });
+        } catch (Exception e) {
+          LOG.warn("Got exception while accessing path {}", uri.toString(), e);
+        }
+      }
+    } while (nextTableKey != null);
+  }
+
+  /**
+   *
+   * @param namespaceId namespace for which metadata needs to be scanned
+   * @param namespacedLogBaseDir namespaced log base dir without the root dir prefixed
+   * @param diskFilesProcessor processor for disk files
+   * @return set of locations on disk
+   */
+  private Set<Location> getDiskLocations(NamespaceId namespaceId, final Location namespacedLogBaseDir,
+                                         final Processor<LocationStatus, Set<Location>> diskFilesProcessor) {
+    try {
+      impersonator.doAs(namespaceId, new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          Locations.processLocations(namespacedLogBaseDir, true, diskFilesProcessor);
+          return null;
+        }
+      });
+    } catch (Exception e) {
+      LOG.warn("Got exception while accessing path {}", namespacedLogBaseDir, e);
+    }
+    return diskFilesProcessor.getResult();
+  }
+
+  private Processor<URI, Set<URI>> getMetaFilesProcessor() {
+    return new Processor<URI, Set<URI>>() {
+      private Set<URI> locations = new HashSet<>();
+
+      @Override
+      public boolean process(final URI inputUri) {
+        locations.add(inputUri);
+        return true;
+      }
+
+      @Override
+      public Set<URI> getResult() {
+        return locations;
+      }
+    };
+  }
+
+  /**
+   * Logging context for a namespace
+   */
+  private class LogNamespaceLoggingContext extends NamespaceLoggingContext {
+    /**
+     * Constructs NamespaceLoggingContext.
+     *
+     * @param namespaceId namespace id
+     */
+    private LogNamespaceLoggingContext(String namespaceId) {
+      super(namespaceId);
+    }
+  }
+
+  private Processor<LocationStatus, Set<Location>> getDiskFilesProcessor(final NamespaceId namespaceId,
+                                                                         final long tillTime) {
+    return new Processor<LocationStatus, Set<Location>>() {
+      private Set<Location> locations = new HashSet<>();
+      private int count = 0;
+
+      @Override
+      public boolean process(final LocationStatus input) {
+        try {
+          Location location = impersonator.doAs(namespaceId, new Callable<Location>() {
+            @Override
+            public Location call() throws Exception {
+              return rootLocationFactory.create(input.getUri());
+            }
+          });
+          // For each namespace, only scan MAX_DISK_FILES_SCANNED disk files per clean up run
+          if (!input.isDir() && location.lastModified() < tillTime && count < MAX_DISK_FILES_SCANNED) {
```

Comment:
When `count` exceeds `MAX_DISK_FILES_SCANNED`, we can return false from the processor so that further processing stops.